tests/test_ssrf_filter_proxy_support.py

```diff
@@ -0,0 +1,156 @@
+"""
+This test file checks that the proxy support is behaving as expected.
+
+``requests`` and ``urllib3`` only support SOCKS4, SOCKS5, and HTTP proxies,
+thus we are only testing these.
+"""
+
+import functools
+
+import pytest
+from pytest import FixtureRequest
+from requests.exceptions import SSLError
+from urllib3.exceptions import MaxRetryError
+from urllib3.util.ssl_match_hostname import CertificateError
+
+from requests_hardened.ip_filter import InvalidIPAddress
+from tests.http_managers import SSRFFilter, SSRFFilterAllowLocalHost
+from tests.http_test_servers import SNITLSHTTPTestServer, TLSTestServer
+from tests.utils import mock_getaddrinfo
+
+# Lists the supported proxy server protocols.
+# These are the ones supported by `requests[socks]` and `urllib3`.
+SUPPORTED_PROXY_PROTOCOLS = ("socks4", "socks5", "http")
+
+
+@pytest.mark.parametrize(
+    # The protocol to use inside the requests.get(...)
+    "http_client_proto",
+    ("http", "https"),
+)
+@pytest.mark.parametrize(
+    # The protocol to use as the proxy server backend (e.g., SOCKS5)
+    "proxy_proto",
+    SUPPORTED_PROXY_PROTOCOLS,
+)
+@pytest.mark.enable_socket  # We need to be able to create the dummy server
+def test_proxy_ip_filter_blocks_private(
+    http_client_proto: str, proxy_proto: str, request: FixtureRequest
+):
+    """Ensures private IP addresses are blocked when using proxies."""
+
+    proxy_url = request.getfixturevalue(f"dummy_proxy_{proxy_proto}")
+    proxies = {"https": proxy_url, "http": proxy_proto}
+
+    # Test: ensure IP filter works when the IP address is resolved through DNS.
+    with mock_getaddrinfo("10.0.0.1"):
+        with pytest.raises(InvalidIPAddress, match="10.0.0.1"):
+            SSRFFilter.send_request(
+                "GET",
+                f"{http_client_proto}://example.test",  # Use hostname (DNS)
+                proxies=proxies,
+            )
+
+    # Test: ensure IP filter works when the IP address is in the URL (no DNS).
+    with pytest.raises(InvalidIPAddress, match="127.100.0.2"):
+        SSRFFilter.send_request(
+            "GET",
+            f"{http_client_proto}://127.100.0.2",  # Use IP address (no DNS)
+            proxies=proxies,
+        )
+
+
+@pytest.mark.parametrize(
+    # The protocol to use as the proxy server backend (e.g., SOCKS5)
+    "proxy_proto",
+    SUPPORTED_PROXY_PROTOCOLS,
+)
+@pytest.mark.fake_resolver(enabled=True)
+@pytest.mark.enable_socket  # We need to be able to create the dummy server
```

Comment:
Instead of fully enabling the socket, wouldn't be good to enable single target address explicitly?

`@pytest.mark.allow_hosts(['127.0.0.1'])` [ref](https://github.com/miketheman/pytest-socket?tab=readme-ov-file#usage)

Reply:
Fixed in 1eb5b4dba16272e91d0f88872dbd7e440137c4e9, thanks!